[quant] Add FP16Observer for fp16 quant support

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42348 [quant] Use PlaceholderObserver instead of Fp16Observer and NoopObserver
* #42222 [quant][graph] Add support for FP16 dynamic quant
* **#42221 [quant] Add FP16Observer for fp16 quant support**
* #42147 [quant] Add saturate_to_fp16 op for FP16 quant support

Summary:
Adds a new observer that emits a warning if the range of tensor is beyond fp16 range. This will be further used in graph mode quantization to insert the cast to fp16 ops in the graph

Test Plan:
python test/test_quantizaton.py TestObserver.test_fp16_observer

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849222](https://our.internmc.facebook.com/intern/diff/D22849222)

supriyar committed:
[quant] Add FP16Observer for fp16 quant support

Summary:
Adds a new observer that emits a warning if the range of tensor is beyond fp16 range. This will be further used in graph mode quantization to insert the cast to fp16 ops in the graph

Test Plan:
python test/test_quantizaton.py TestObserver.test_fp16_observer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add FP16Observer for fp16 quant support"


Summary:
Adds a new observer that emits a warning if the range of tensor is beyond fp16 range. This will be further used in graph mode quantization to insert the cast to fp16 ops in the graph

Test Plan:
python test/test_quantizaton.py TestObserver.test_fp16_observer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/observer.py:
```diff
@@ -985,6 +985,36 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
         super(HistogramObserver, self)._load_from_state_dict(state_dict, prefix, local_metadata, strict,
                                                              missing_keys, unexpected_keys, error_msgs)
 
+class Fp16Observer(ObserverBase):
```
this seems to be the same as min max observer except for the warning, can we just use min max observer and put warning somewhere else?

supriyar replied:
We don't have calculate_qparams and other methods defined for this observer. Also, feel that from user standpoint it might be better to separate this from MinMax observer since that actually observes the values to calculate the qparams. I can add a docblock to better explain the usage of this.

jerryzh168 replied:
I see, since we just use min_val/max_val for warning, can we remove it?

jerryzh168 replied:
If we can remove the warning then we can just use NoOp observer for this

supriyar replied:
Having the warning is useful for users to know if their weight values might be getting saturated and it can tell them that they might need to update the model. So users can just insert the observers and run the model to check for potential overflow issues.

jerryzh168 replied:
can this be done during convert though? we only have one user facing API quantize_dynamic_jit

supriyar committed:
Update on "[quant] Add FP16Observer for fp16 quant support"


Summary:
Adds a new observer that emits a warning if the range of tensor is beyond fp16 range. This will be further used in graph mode quantization to insert the cast to fp16 ops in the graph

Test Plan:
python test/test_quantizaton.py TestObserver.test_fp16_observer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add FP16Observer for fp16 quant support"


Summary:
Adds a new observer that emits a warning if the range of tensor is beyond fp16 range. This will be further used in graph mode quantization to insert the cast to fp16 ops in the graph

Test Plan:
python test/test_quantizaton.py TestObserver.test_fp16_observer

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849222](https://our.internmc.facebook.com/intern/diff/D22849222)

[ghstack-poisoned]
